### PR TITLE
GP-16134 SqlTasks Action Templates

### PR DESCRIPTION
**Pull request by viliam-ivanov**
**new table civicrm_sqltasks_action_template** (de.systopia.sqltasks/xml/schema/CRM/Sqltasks/)
SqltasksActionTemplates.EntityType.xml
SqltasksActionTemplates.xml

**new ng directives** (de.systopia.sqltasks/ang/sqlTaskConfigurator/)
actionTemplateSelect.html
actionTemplateForm.html

**tests for new Api**
`phpunit5 tests/phpunit/api/v3/SqlTasksActionTemplate/`

### CODE REVIEW

File: CRM/Sqltasks/BAO/SqltasksActionTemplate.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
+<?php
+use CRM_Sqltasks_ExtensionUtil as E;
+
+class CRM_Sqltasks_BAO_SqltasksActionTemplate extends CRM_Sqltasks_DAO_SqltasksActionTemplate {
```
Comment (pfigel) [inline CRM/Sqltasks/BAO/SqltasksActionTemplate.php:4]: Most of the methods in this class are used in the corresponding API3 code. It's generally not necessary to write this low-level code as `civix generate:entity` with `--api-version 3` (or `--api-version 3,4`) will auto-generated API functions/classes for all CRUD operations. The generated code for API3 may look similar to [this](https://github.com/greenpeace-cee/dialfire/blob/main/api/v3/DialfireCampaign.php).

We can probably remove most of this code and the custom API implementation files (i.e. `api/v3/SqltasksActionTemplate/Create.php` and others) if you re-run `civix generate:entity SqltasksActionTemplate --api-version 3`. The [documentation](https://docs.civicrm.org/dev/en/latest/extensions/civix/#generate-entity) describes this process quite well.

Note: Please see my comments re: the schema files below before doing that.

File: xml/schema/CRM/Sqltasks/SqltasksActionTemplates.xml [hunk 2]
```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="iso-8859-1" ?>
```
Comment (pfigel) [inline xml/schema/CRM/Sqltasks/SqltasksActionTemplates.xml:1]: The file name uses plural here; the convention in Civi is to use singular, plus `civix` and other parts of Civi expect schema files and their corresponding classes to have matching names.

File: xml/schema/CRM/Sqltasks/SqltasksActionTemplates.entityType.php [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?php
```
Comment (pfigel) [inline xml/schema/CRM/Sqltasks/SqltasksActionTemplates.entityType.php:1]: Same singular/plural issue as `xml/schema/CRM/Sqltasks/SqltasksActionTemplates.xml`.

File: xml/schema/CRM/Sqltasks/SqltasksActionTemplates.xml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+<table>
+  <base>CRM/Sqltasks</base>
+  <class>SqltasksActionTemplates</class>
```
Comment (pfigel) [inline xml/schema/CRM/Sqltasks/SqltasksActionTemplates.xml:5]: Should also be singular to match the DAO/BAO class name.

File: CRM/Sqltasks/BAO/SqltasksActionTemplate.php (hunk 1 above)
Comment (viliam-ivanov) [reply CRM/Sqltasks/BAO/SqltasksActionTemplate.php:4]: Resolved with commit 1d6fc5bd5f0ea6a622db81cccb2fac28b02da05c [Updated api calls and result values sequential]

**Review by mflandorfer — APPROVED** [review]

File: CRM/Sqltasks/Upgrader.php [hunk 5]
```diff
@@ -383,4 +383,27 @@ public function upgrade_0150 () {
...
+      $logging = new CRM_Logging_Schema();
+      $logging->fixSchemaDifferences();
+    }
+  }
+
+  /**
+   * @return bool
+   * @throws \Exception
+   */
+  public function upgrade_200 () {
```
[the remark this answers is not in the document]
Comment (pfigel) [reply CRM/Sqltasks/Upgrader.php:405]: yes, that makes more sense.

File: ang/sqlTaskConfigurator.js [hunk 6]
```diff
@@ -984,13 +1003,24 @@
     return false;
   }
 
+  function toggleActionTemplateForm() {
```
[the remark this answers is not in the document]
Comment (viliam-ivanov) [reply ang/sqlTaskConfigurator.js]: Resolved with commit e3517c006d4bb621b0ae0037677e04f14ea4e973

File: CRM/Sqltasks/BAO/SqltasksActionTemplate.php [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+  /**
+   * Create a new SqltasksActionTemplate based on array-data
+   *
+   * @param array $params key-value pairs
+   * @return CRM_Sqltasks_DAO_SqltasksActionTemplate|NULL
+   *
+  public static function create($params) {
```
[the remark this answers is not in the document]
Comment (viliam-ivanov) [reply CRM/Sqltasks/BAO/SqltasksActionTemplate.php]: Resolved with commit https://github.com/greenpeace-cee/de.systopia.sqltasks/commit/e3517c006d4bb621b0ae0037677e04f14ea4e973

File: tests/phpunit/api/v3/SqlTasksActionTemplate/CreateTest.php [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+  public function tearDown() {
+    parent::tearDown();
+  }
+
+  /**
+   * Test creation of a template
+   */
+  public function testCreateTemplate() {
+    try {
+      $templatesCountBefore = civicrm_api3('SqltasksActionTemplate', 'get')["count"];
```
[the remark this answers is not in the document]
Comment (viliam-ivanov) [reply tests/phpunit/api/v3/SqlTasksActionTemplate/CreateTest.php]: Resolved with commit https://github.com/greenpeace-cee/de.systopia.sqltasks/commit/e3517c006d4bb621b0ae0037677e04f14ea4e973

File: CRM/Sqltasks/Upgrader.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (viliam-ivanov) [reply CRM/Sqltasks/Upgrader.php]: Resolved with commit https://github.com/greenpeace-cee/de.systopia.sqltasks/commit/e3517c006d4bb621b0ae0037677e04f14ea4e973

**Review by mflandorfer — APPROVED** [review]

**Pull request merged**